docs/source/developer_guides/model_merging.md

```diff
@@ -0,0 +1,136 @@
+
+Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+specific language governing permissions and limitations under the License.
+
+⚠️ Note that this file is in Markdown but contain specific syntax for our doc-builder (similar to MDX) that may not be
+rendered properly in your Markdown viewer.
+
+-->
+
+# Model merging
+
+Training a model for each task can be costly, take up storage space, and the models aren't able to learn new information to improve their performance. Multitask learning can overcome some of these limitations by training a model to learn several tasks, but it is expensive to train and designing a dataset for it is challenging. *Model merging* offers a solution to these challenges by combining multiple pretrained models into one model, giving it the combined abilities of each individual model without any additional training.
+
+PEFT provides two methods for model merging:
+
+* [TIES](https://hf.co/papers/2306.01708) - TrIm, Elect, and Merge (TIES) is a three-step method for merging models. First, redundant parameters are trimmed, then conflicting signs are resolved into an aggregated vector, and finally the parameters whose signs are the same as the aggregate sign are averaged. This method takes into account that some values (redundant and sign disagreement) can degrade performance in the merged model.
+* [DARE](https://hf.co/papers/2311.03099) - Drop And REscale is a method that can be used to prepare for other model merging methods like TIES. It works by randomly dropping parameters according to a drop rate and rescaling the remaining parameters. This helps to reduce the number of redundant and potentially interfering parameters among multiple models.
+
+Models are merged with the [`~LoraModel.add_weighted_adapter`] method, and the specific model merging method is specified in the `combination_type` parameter. This guide will show you how to merge models with TIES and DARE.
+
+## Merge method
+
+With PEFT, merging is enabled by setting `combination_type` and `ties_density` to a value of the weights to keep from the individual models. For example, let's merge three finetuned [TinyLlama/TinyLlama-1.1B-intermediate-step-1431k-3T](https://huggingface.co/TinyLlama/TinyLlama-1.1B-intermediate-step-1431k-3T) models: [tinyllama_lora_nobots](https://huggingface.co/smangrul/tinyllama_lora_norobots), [tinyllama_lora_sql](https://huggingface.co/smangrul/tinyllama_lora_sql), and [tinyllama_lora_adcopy](https://huggingface.co/smangrul/tinyllama_lora_adcopy).
+
+<Tip>
+
+If you're merging with TIES, use the [`~transformers.PreTrainedModel.resize_token_embeddings`] method to account for special tokens added to the embedding layer for each finetuned model. This method ensures the special tokens and initialization of the embedding layers are consistent.
+
+</Tip>
```

Comment:
I am not sure if I understand this correctly. Just to clarify what I meant in my earlier comment, if we are merging say three full Models like, WizardMath, WizardCode which are based on the common LLaMA checkpoints which has vocab of say 32k. Say when finetuning on LLaMA model on code data to get WizardCode model some new special tokens might be added which are not in the LLaMA tokenizer, say "{", "}" are the two special tokens added. Similarly for WizardMath we might have added say "=" as a special token.  Now in the embedding layer of WizardCode the 32001 embedding correspond to "{" and 32002 corresponds to "}". However, in WizardMath the 32001 embedding corresponds to "=". Hence, we need to merge the embedding layer such that we do not merge the learned representation of "{" from WizardCode with "=" from WizardMath which both have index 32001. 

However, this issue can only occur when merging full model and should not be an issue when merging PEFT modules trained on a shared base model. I am not sure if this is clear from the tip that has been added. It is also possible that I am misunderstanding something.

Reply:
Thanks, I tried to clarify it a bit more based on your comment